socorro/external/boto/connection_context.py

```diff
@@ -47,280 +81,135 @@ class S3ConnectionContext(RequiredConfig):
-        from_string_converter=class_converter,
-        reference_value_from="resource.boto",
-        likely_to_be_changed=True,
-    )
 
-    RETRYABLE_EXCEPTIONS = (
-        socket.timeout,
-        boto.exception.PleaseRetryException,
-        boto.exception.ResumableUploadException,
-    )
+    KeyNotFound = KeyNotFound
 
-    def __init__(self, config, quit_check_callback=None):
+    def __init__(self, config):
         self.config = config
-        self._CreateError = boto.exception.StorageCreateError
-        self.ResponseError = (boto.exception.StorageResponseError, KeyNotFound)
-        self._bucket_cache = {}
-        self.metrics = markus.get_metrics(config.boto_metrics_prefix)
-        self._connect_to_endpoint = boto.connect_s3
-        self._calling_format = config.calling_format
-
-    def _connect(self):
-        try:
-            return self.connection
-        except AttributeError:
-            self.connection = self._connect_to_endpoint(**self._get_credentials())
-            return self.connection
-
-    def _get_credentials(self):
-        return {
-            "aws_access_key_id": self.config.access_key,
-            "aws_secret_access_key": self.config.secret_access_key,
-            "is_secure": True,
-            "calling_format": self._calling_format(),
+        self.bucket = self.config.bucket_name
+        self.client = self.build_client()
+
+    @retry(
+        retryable_exceptions=[
+            # FIXME(willkg): Seems like botocore always raises ClientError
+            # which is unhelpful for granularity purposes.
+            ClientError,
+            # This raises a ValueError "invalid endpoint" if it has problems
+            # getting the s3 credentials and then tries "s3..amazonaws.com"--we
+            # want to retry that, too.
+            ValueError,
+        ],
+        wait_time_generator=wait_times_connect,
+        sleep_function=time.sleep,
+        module_logger=logger,
+    )
+    def build_client(self):
+        """Returns a Boto3 S3 Client."""
+        # Either they provided ACCESS_KEY and SECRET_ACCESS_KEY in which case
+        # we use those, or they didn't in which case boto3 pulls credentials
+        # from one of a myriad of other places.
+        # http://boto3.readthedocs.io/en/latest/guide/configuration.html#configuring-credentials
+        session_kwargs = {}
+        if self.config.access_key and self.config.secret_access_key:
+            session_kwargs["aws_access_key_id"] = self.config.access_key
+            session_kwargs["aws_secret_access_key"] = self.config.secret_access_key
+        session = boto3.session.Session(**session_kwargs)
+
+        kwargs = {
+            "service_name": "s3",
+            "region_name": self.config.region,
+            # NOTE(willkg): We use path-style because that lets us have dots in
+            # our bucket names and use SSL.
+            "config": Config(s3={"addressing_style": "path"}),
```

Comment:
Is the change of addressing style needed? Other comments suggest that we no longer use periods in bucket names.

Reply:
I forgot about this. I did write comments elsewhere about this. I'll have to double-check to see what the deal is, but I'm pretty sure we're fine now.

If we change this code, I'll make sure to change Antenna to match. At a minimum, I'll update the comment.

Reply:
This line was in Antenna from the beginning (September 2016) and at that time, we did have periods in the bucket names. Pretty sure we don't anymore because periods are the worst.

"virtual" `addressing_style` also doesn't work with the local dev environment because it starts building paths like `http://dev_bucket.localstack-s3:3000/` and that won't work.

However, it looks like the default for addressing_style is "auto" where it'll try "virtual' first and if that fails, use "path":

https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html?highlight=addressing_style#configuration-file

Looks like that was changed in botocore 1.9.0 in 2018:

* https://github.com/boto/botocore/blob/develop/CHANGELOG.rst#190
* https://github.com/boto/botocore/pull/1387

I'll remove these three lines from Socorro and Antenna.